tests/test_analysis.py

```diff
@@ -637,7 +637,31 @@ def test_summarize():
         pd.DataFrame(
             columns=pd.Index(["lower", "mean", "upper"], name="stat"),
             index=["TimePoint0", "TimePoint1"],
-            data=np.array([[0.5, 10.0, 19.5], [0.5, 10.0, 19.5],]),
+            data=np.array([[0.5, 10.0, 19.5], [0.5, 10.0, 19.5], ]),
         ),
         summarize(results_one_draw, collapse_columns=True),
     )
+
+
+def test_auto_register_modules(tmpdir):
+    """ check module dependencies can be registered automatically """
+    start_date = Date(2010, 1, 1)
+    # configure logging
+    log_config = {
+        "filename": "LogFile",
+        "directory": tmpdir,
+    }
+    sim = Simulation(start_date=start_date, seed=0, log_config=log_config, data_folder=resourcefilepath)
+    try:
+        # try executing the code in this block and go to except block if module dependency error exception is fired
+
+        # register modules without their associated dependencies
+        sim.register(demography.Demography(resourcefilepath=resourcefilepath),
+                     copd.Copd(resourcefilepath=resourcefilepath),
+                     auto_register_modules=True)
+
+    except ModuleDependencyError as exception:
+        # if auto register modules argument is false, there should be a module dependency error exception fired
+        assert exception
+        assert exception.__class__ == ModuleDependencyError
+
```

Comment:
Thanks @matt-graham. Part of your comment on the test is what also was also going on in my mind(I wasn't fully satisfied with it). Initially my plan was to use `with pytest.raises(ModuleDependencyError)`. The only challenge is that this fails when the code is not raising any exception(with an error `DID NOT RAISE`) and I wasn't able to find a way how I can capture that(as in how to assert that no exception was raised)

Reply:
on where to place this test. I think test dependencies will also be fine. I just noted that there is already a test in there(`test_missing_dependency_raises_error_on_register`, using dummy modules of course) that checks module dependency and I felt like I should  not add another test that to some extent will do the same. Do you want me to extend/update this test with real disease modules?

Reply:
Ah okay - for testing that a code snippet doesn't raise an exception the usual pattern is to just have a test function which calls the relevant function / methods, as if an unhandled exception is encountered this will automatically be interpreted as a test failure. Your current test with the `sim.register` call moved outside of the `try` block (and no corresponding `except` block) would work for this. While testing no error is raised is a useful test, I would also say it would be worth testing if no error happens whether the registered dependencies are as expected using something similar to what I suggested above. The test should then fail either if we get a module dependency error unexpectedly or if the module dependencies are not as expected (but we didn't get an error).

Reply:
So helpful. Thanks @matt-graham 

> Ah okay - for testing that a code snippet doesn't raise an exception the usual pattern is to just have a test function which calls the relevant function / methods, as if an unhandled exception is encountered this will automatically be interpreted as a test failure. Your current test with the `sim.register` call moved outside of the `try` block (and no corresponding `except` block) would work for this. While testing no error is raised is a useful test, I would also say it would be worth testing if no error happens whether the registered dependencies are as expected using something similar to what I suggested above. The test should then fail either if we get a module dependency error unexpectedly or if the module dependencies are not as expected (but we didn't get an error).